### ISSUE TITLE
[REVIEW]: A Framework to Quality Control Oceanographic Data

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2063]
**Submitting author:** @castelao (<a href="http://orcid.org/0000-0002-6765-0708">Guilherme Castelao</a>)
**Repository:** <a href="https://github.com/castelao/CoTeDe" target ="_blank">https://github.com/castelao/CoTeDe</a>
**Version:** v0.21.3
**Editor:** @kthyng
**Reviewer:** @jessicaaustin, @evanleeturner
**Archive:** <a href="https://doi.org/10.5281/zenodo.3733959" target="_blank">10.5281/zenodo.3733959</a>

## Status

[![status](https://joss.theoj.org/papers/87191fa86135614f71cc945c36b5f532/status.svg)](https://joss.theoj.org/papers/87191fa86135614f71cc945c36b5f532)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/87191fa86135614f71cc945c36b5f532"><img src="https://joss.theoj.org/papers/87191fa86135614f71cc945c36b5f532/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/87191fa86135614f71cc945c36b5f532/status.svg)](https://joss.theoj.org/papers/87191fa86135614f71cc945c36b5f532)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/castelao/CoTeDe" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@jessicaaustin & @evanleeturner, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @kthyng know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/castelao/CoTeDe">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@castelao) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/castelao/CoTeDe">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@castelao) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @jessicaaustin, @evanleeturner it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1016/j.mio.2014.09.001 is OK

MISSING DOIs

- None

INVALID DOIs

-  http://dx.doi.org/10.13155/33951 is INVALID because of 'https://doi.org/' prefix
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02063/joss.02063/10.21105.joss.02063.pdf)

**Comment by castelao** [comment]
Hi everyone. It's three weeks since the start of the review. Would you have any update for me? Thanks.

**Comment by kthyng** [comment]
I'll just do a specific ping to reviewers @jessicaaustin and @evanleeturner. When do you think you'll be able to work on your reviews? Thanks.

**Comment by evanleeturner** [comment]
I am still working on this review, thank you.

**Comment by jessicaaustin** [comment]
Thanks for the reminder. I have read all the instructions and made some progress on my review before getting pulled into other tasks. I will pick it up again this week and submit by the end of the week.

**Comment by jessicaaustin** [comment]
### Conflict of interest

I do not believe this is a conflict of interest as outlined in the JOSS guidelines, but I do want to state that I am a maintainer of a similar open-source QC library, https://github.com/ioos/ioos_qc .

### Overall thoughts:

I agree 100% with the author's philosophy that QC procedures should be organized and distributed as code packages in order to widely and consistently used. I also agree with the goal of creating a framework that can run tests from multiple different QC test schemes (GTSPP, ARGO, QARTOD, etc) since none of those schemes alone will completely satisfy testing requirements in all situations. I think the integration with oceansdb to run Climatology tests and Location at Sea tests is particularly exciting. I hope publication in JOSS will bring this library to the attention of more groups and encourage further use and collaboration to flesh this out.

The library takes in data as a numpy array, so in theory the data could come from any source or format. However you must create a "data object" to wrap your data. Given that there are other mature and well-used libraries with data representations (xarray, pandas, netcdf, etc), why not leverage one of those? Or at least make it easy to use CoTeDe if your data is already in one of those common formats. I created a notebook to try to understand how to run QC tests using the library ( https://github.com/castelao/CoTeDe/pull/42), and I was able to fairly easily load a netcdf file using xarray and run tests, but I got an error when using a pandas dataframe and it was not clear how to debug it. I think a little documentation could go a long way here, to help people get started.

### Review

Items marked **BLOCKER** below should be resolved before acceptance. Everything else is a recommendation.

Paper:

* Overall this paper is clear and easy to read, and does a good job of describing the motivation and overall functionality of the software
* **Highly recommended:** It would be great to list similar automated QC packages and describe how CoTeDe compares. For example, CoTeDe is more generic than ioos_qc, which depends on pandas/xarray and assumes netCDF integration, and CoTeDe allows integration with a wide range of QC test suites, whereas ioos_qc (currently) only focuses on QARTOD.  
   * I do understand that this takes time, so I don't want it to block acceptance here. But this is the first question I had when I looked at this library, and would love to see it in the README someday.
* Highly recommended, if possible: Include a list of groups actively using this library, to give us an idea of how widely used it is. (this could also be added to the code documentation)
* Recommended: Describe how to pronounce CoTeDe (though I suppose leaving some amount of mystery is the python way :-) )


Documentation:

* **BLOCKER:** The important classes and most of the test methods have docstrings, which is great. But I was not able to find any generated API docs under https://cotede.readthedocs.io/ . https://github.com/castelao/CoTeDe/issues/41


Installation

* I was able to install and use the software using the instructions in the repo
* Nice to have: https://github.com/castelao/CoTeDe/issues/39

Contributing

* I was able to install dependencies and run tests locally using the instructions in the repo
* Recommended: https://github.com/castelao/CoTeDe/pull/38
* Recommended: https://github.com/castelao/CoTeDe/issues/40

Performance

* I did not find any mention of performance testing or guarantees in the code or documentation
* Highly recommended you add performance tests. During development of https://github.com/ioos/ioos_qc we found that small tweaks to our testing algorithms could lead to severely degraded performance on large datasets. Recommend you add these tests to the travis-ci build and add it to the checklist of items to review if someone submits a PR.

**Comment by castelao** [comment]
@jessicaaustin , thank you for your review. I'll respond to it ASAP.

**Comment by evanleeturner** [comment]
### Conflict of interest

I do not have a conflict of interest in this review.  I do maintain internal (non-public) code repository @wdft-coastal that does contain some of the similar functions as presented in the current code but are specific for my dataset and organization.  

### Overall thoughts:
I am extremely pleased with the premise of this package for organizing QC procedures and being able to run multiple tests.  This is a high priority item for my organization and my current data sets, as our own QAQC methods often can conflict with other agencies that follow different QAQC procedures. In the last few months I have been preparing code that can mimic QARTOD style tests on our own data internally, but I wanted the ability to apply alternative test schemes and to skip tests that are not appropriate for our data.  Having a maintained package that accomplishes this task is not only helpful but it also greatly increases your data confidence as now the QC methods are maintained by other groups in a transparent mechanism.  I fully expect to be implementing this package into our own website and data products at TWDB. 
 
I share reviewer one’s sentiment about the philosophy of leveraging other data libraries such as netCDF or pandas.  Although our own internal datasets are SQL we do make use of pandas where possible for data manipulation and in the future we will likely be forced to move to netCDF to process very large datasets for QAQC.  This functionality is not a show-stopper for the package, but it is a severe limitation for its acceptance in the community.

### Review
Items marked **BLOCKER** below should be resolved before acceptance. Everything else is a recommendation.
Paper:
•	The paper itself was succinct, well-written, and very appropriate to the code.
•	**Recommended:**  Please consider adding the functionality of swapping methods for calculating seawater: https://github.com/castelao/CoTeDe/issues/43 
•	**Recommended:** mirror reviewer one that I don’t know how to pronounce CoTeDe! 😊
Documentation:
•	**BLOCKER**: Please also complete the issue that was brought by reviewer one concerning the API docs as this would also be extremely helpful for me to implement this code: https://cotede.readthedocs.io/ . castelao/CoTeDe#41
Installation
•	I was able to get the package running.
Contributing
•	I was able to run the tests.
Performance
•	There were no performance tests or statements regarding performance. 
•	I did not have an issue with performance with my own tests, but I also have very small datasets at the moment.  Additionally, if I were to run this package in a production environment, the performance hit would be negated because our data would be checked as it inserted into our database which would be a heftier cost of CPU cycles.  If I were going to run these tests on the entire database, I would do so offline or during non-peak hours on the production systems in any case.

**Comment by castelao** [comment]
@evanleeturner, thanks for the review. I already started to work on it.

**Comment by castelao** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02063/joss.02063/10.21105.joss.02063.pdf)

**Comment by castelao** [comment]
Hi @evanleeturner , you might be able to edit the top post, and if so, could you click on the review items that you agree with, please? That would help me to write my response. Thanks!

**Comment by castelao** [comment]
Response to @jessicaaustin 

> ### Conflict of interest
> 
> I do not believe this is a conflict of interest as outlined in the JOSS guidelines, but I do want to state that I am a maintainer of a similar open-source QC library, https://github.com/ioos/ioos_qc .
> ### Overall thoughts:
> 
> I agree 100% with the author's philosophy that QC procedures should be organized and distributed as code packages in order to widely and consistently used. I also agree with the goal of creating a framework that can run tests from multiple different QC test schemes (GTSPP, ARGO, QARTOD, etc) since none of those schemes alone will completely satisfy testing requirements in all situations. I think the integration with oceansdb to run Climatology tests and Location at Sea tests is particularly exciting. I hope publication in JOSS will bring this library to the attention of more groups and encourage further use and collaboration to flesh this out.

Thanks! As a side note, the package OceansDB was originally together with CoTeDe, as a  single package (since the first generation, ~ 2006), but as CoTeDe grew larger and I realized that some people could use OceansDB outside CoTeDe’s scope, I moved it outside as an independent package (~ 2015).

> The library takes in data as a numpy array, so in theory the data could come from any source or format. However you must create a "data object" to wrap your data. Given that there are other mature and well-used libraries with data representations (xarray, pandas, netcdf, etc), why not leverage one of those? Or at least make it easy to use CoTeDe if your data is already in one of those common formats. I created a notebook to try to understand how to run QC tests using the library ( [castelao/CoTeDe#42](https://github.com/castelao/CoTeDe/pull/42)), and I was able to fairly easily load a netcdf file using xarray and run tests, but I got an error when using a pandas dataframe and it was not clear how to debug it. I think a little documentation could go a long way here, to help people get started.

Thanks for sharing your opinion on using xarray or pandas. It’s now clear for me that I need to clarify this in the documentation. This will be addressed at https://cotede.readthedocs.io/en/dev/data_model.html

In summary, I believe that the data model adopted by CoTeDe does not limit the users, but the other way around: it gives more freedom. The package xarray, which was initially based on the netCDF structure, satisfies the data model expected by CoTeDe, thus most of CoTeDe’s functionalities can be used without any change, [see the example](https://github.com/castelao/CoTeDe/blob/master/docs/notebooks/TSG_Saildrone.ipynb). Since xarray can read directly netCDF and pandas objects can be converted into xarray objects, all those inputs can be used by CoTeDe, plus any other format that satisfies CoTeDe’s minimalist data model. CoTeDe has been used with SQL databases and MatLab based pipelines, just two examples where less requirements makes a difference.

> ### Review
> 
> Items marked **BLOCKER** below should be resolved before acceptance. Everything else is a recommendation.
> 
> Paper:
> 
>     * Overall this paper is clear and easy to read, and does a good job of describing the motivation and overall functionality of the software
> 
>     * **Highly recommended:** It would be great to list similar automated QC packages and describe how CoTeDe compares. For example, CoTeDe is more generic than ioos_qc, which depends on pandas/xarray and assumes netCDF integration, and CoTeDe allows integration with a wide range of QC test suites, whereas ioos_qc (currently) only focuses on QARTOD.
>       
>       * I do understand that this takes time, so I don't want it to block acceptance here. But this is the first question I had when I looked at this library, and would love to see it in the README someday.

I see your point and I will consider that. The brief comparison that you suggested above could be a good start, but what should be said about the similarities between the two packages? For instance, what should we say about QC configuration, which seems like ioss_qc followed a quite similar approach as the one proposed by CoTeDe. Would you agree? On a side note, since this is a highly recommended suggestion, I was expecting that ioos_qc would adopt the same policy, but I could not find any reference to CoTeDe in the ioos_qc documentation.

>     * Highly recommended, if possible: Include a list of groups actively using this library, to give us an idea of how widely used it is. (this could also be added to the code documentation)

Thanks, that’s a good idea. However, I believe that the choice to be listed belongs to the users. I’ll create a designated space for that on CoTeDe’s website and encourage the users to add themselves to the list.

>     * Recommended: Describe how to pronounce CoTeDe (though I suppose leaving some amount of mystery is the python way :-) )

At some point I removed that from the documentation, but I’ll place it back.

> 
> Documentation:
> 
>     * **BLOCKER:** The important classes and most of the test methods have docstrings, which is great. But I was not able to find any generated API docs under https://cotede.readthedocs.io/ . [castelao/CoTeDe#41](https://github.com/castelao/CoTeDe/issues/41)

I have expanded the documentation (https://cotede.readthedocs.io/) to include the API reference for the public resources, i.e. everything that a new user would need to know to use CoTeDe. I’m also expanding the notebooks with practical examples to better illustrate how to use the package. I intend to extend the API reference with internal resources that might be useful for developers (https://cotede.readthedocs.io/en/latest/api.html).

> Installation
> 
>     * I was able to install and use the software using the instructions in the repo
> 
>     * Nice to have: [castelao/CoTeDe#39](https://github.com/castelao/CoTeDe/issues/39)

Thanks for the suggestion. I understand the benefits of source-forge and I’ll definitely add CoTeDe into that in the future, but it is not the priority at the moment. I noticed that ioos_qc uses pip on its notebooks, which is yet another example of how PIP is convenient and functional.

> 
> Contributing
> 
>     * I was able to install dependencies and run tests locally using the instructions in the repo
> 
>     * Recommended: [castelao/CoTeDe#38](https://github.com/castelao/CoTeDe/pull/38)
> 
>     * Recommended: [castelao/CoTeDe#40](https://github.com/castelao/CoTeDe/issues/40)

I removed the recommendation on flake8. I do not want to turn back any potential contribution just due to formatting. I’ve been cleaning, refactoring, and formatting the code and I will continue on doing so.

> 
> Performance
> 
>     * I did not find any mention of performance testing or guarantees in the code or documentation
> 
>     * Highly recommended you add performance tests. During development of https://github.com/ioos/ioos_qc we found that small tweaks to our testing algorithms could lead to severely degraded performance on large datasets. Recommend you add these tests to the travis-ci build and add it to the checklist of items to review if someone submits a PR.

Thank you, that is a very good idea. I've always assumed that by using Python it was implied that speed was not a priority, but there is nothing to lose on having an idea about the relative cost among the different tests. I’ll implement that on the first chance. Note that one of the consistency tests is to guarantee that the ProfileQC object is serializable, so that it can be transported by multiprocessing. On large scale databases I have been running CoTeDe in parallel, a functionality that I would like to move to the public repository on the first chance.

A note, when I read "performance" for a QC procedure the first thing that comes to my mind is not speed but the skill of identifying bad measurements without mistakes. For that, I’ve been adopting [consistency tests](https://github.com/castelao/CoTeDe/tree/master/tests) since the early stages of CoTeDe.

**Comment by castelao** [comment]
Response to @evanleeturner 

> ### Conflict of interest
> 
> I do not have a conflict of interest in this review. I do maintain internal (non-public) code repository @wdft-coastal that does contain some of the similar functions as presented in the current code but are specific for my dataset and organization.
> ### Overall thoughts:
> 
> I am extremely pleased with the premise of this package for organizing QC procedures and being able to run multiple tests. This is a high priority item for my organization and my current data sets, as our own QAQC methods often can conflict with other agencies that follow different QAQC procedures. In the last few months I have been preparing code that can mimic QARTOD style tests on our own data internally, but I wanted the ability to apply alternative test schemes and to skip tests that are not appropriate for our data. Having a maintained package that accomplishes this task is not only helpful but it also greatly increases your data confidence as now the QC methods are maintained by other groups in a transparent mechanism. I fully expect to be implementing this package into our own website and data products at TWDB.

Thanks, I’m glad to hear that we have the same vision on shared public QC procedures and the importance of giving freedom to the operator to decide which tests to apply, including the tunning of each test. Please, don’t hesitate to open issues with requests or discussions that could help you to implement CoTeDe in your project. My vision for the QC configuration was on cases exactly like yours. I’ll improve the documentation on that subject.
 
> I share reviewer one’s sentiment about the philosophy of leveraging other data libraries such as netCDF or pandas. Although our own internal datasets are SQL we do make use of pandas where possible for data manipulation and in the future we will likely be forced to move to netCDF to process very large datasets for QAQC. This functionality is not a show-stopper for the package, but it is a severe limitation for its acceptance in the community.

Thanks for sharing your opinion on using netCDF or pandas. It’s now clear for me that I need to clarify this in the documentation. This will be addressed at https://cotede.readthedocs.io/en/dev/data_model.html
In summary, I believe that the data model adopted by CoTeDe does not limit the users, but the other way around: it gives more freedom. The package xarray, which was initially based on the netCDF structure, satisfies the data model expected by CoTeDe, thus most of CoTeDe’s functionalities can be used without any change, [see the example:](https://github.com/castelao/CoTeDe/blob/master/docs/notebooks/TSG_Saildrone.ipynb). Since xarray can read directly netCDF and pandas objects can be converted into xarray objects, all those inputs can be used by CoTeDe as it is now, plus any other format that satisfies CoTeDe’s minimalist data model. I also use CoTeDe on a PostgreSQL DB with more than a TB of data, and instead of pandas.read_sql I use psycopg2 directly. I know that CoTeDe has been also used in a MatLab data pipeline, another example where fewer dependencies facilitate its use, especially for MatLab users without Python experience. I believe that this is a case where less is more, and for CoTeDe there is no need to force dependencies on more packages.

> ### Review
> 
> Items marked **BLOCKER** below should be resolved before acceptance. Everything else is a recommendation.
> 
> Paper
> 
>     * The paper itself was succinct, well-written, and very appropriate to the code.
> 
>     * **Recommended:**  Please consider adding the functionality of swapping methods for calculating seawater: [castelao/CoTeDe#43](https://github.com/castelao/CoTeDe/issues/43)

Issue #43 raises a very good point. We will certainly find derived variables (salinity, density, sound speed, etc.) estimated with different standards (especially on historical data), which may affect inter-comparisons. That is indeed important but is beyond CoTeDe’s goal, which is intended to evaluate the quality of a dataset. Currently, the only place that I use GSW is to estimate sea water density when density is not directly available to evaluate inversions in the water column. For that purpose, I argue that we should use the best and most recent technique. I do see your point that in some situations it might be useful, if not necessary, to employ the old relations for a correct comparison, but I think that this should be done outside CoTeDe, by the operator, and then used as input for CoTeDe.
Note that I’m strongly considering adopting your suggestion in another package [PySeabird](https://github.com/castelao/seabird), which is meant to parse and estimate derived variables from CTDs.

>     * **Recommended:** mirror reviewer one that I don’t know how to pronounce CoTeDe! 😊

At some point, I removed that from the documentation, but I’ll put it back.

> Documentation:
> 
>     * **BLOCKER**: Please also complete the issue that was brought by reviewer one concerning the API docs as this would also be extremely helpful for me to implement this code: https://cotede.readthedocs.io/ . [castelao/CoTeDe#41](https://github.com/castelao/CoTeDe/issues/41)

I have expanded the documentation (https://cotede.readthedocs.io/) to include the API reference for the public resources, i.e. everything that a new user would need to know to use CoTeDe. I’m also expanding the notebooks with practical examples to better illustrate how to use the package. I intend to extend the API reference with internal resources, which might be useful for developers (https://cotede.readthedocs.io/en/dev/api.html).

> Installation
> 
>     * I was able to get the package running.
> 
> 
> Contributing
> 
>     * I was able to run the tests.
> 
> 
> Performance
> 
>     * There were no performance tests or statements regarding performance.
> 
>     * I did not have an issue with performance with my own tests, but I also have very small datasets at the moment.  Additionally, if I were to run this package in a production environment, the performance hit would be negated because our data would be checked as it inserted into our database which would be a heftier cost of CPU cycles.  If I were going to run these tests on the entire database, I would do so offline or during non-peak hours on the production systems in any case.

Thanks for raising that point. I’m adding a section in the documentation about the performance. (https://cotede.readthedocs.io/en/dev/performance.html).
I agree with you, the QC in the inserting phase of the pipeline could cause a bottleneck and propagate serious issues. Something I have done before was to implement a parallel process that would run the QC on the data already inserted in the DB, on-demand, triggered by new inserts. The system that I developed for AOML, back in 2006, loaded the Python QC tests as a PostgreSQL procedural language, so it was an internal procedure with an internal trigger, per profile (Again, easier if require less external packages). That probably results in the best speed for a SQL DB, but it is not necessarily the easiest approach to maintain.

**Comment by castelao** [comment]
@kthyng , could you confirm the next step, please? Is it correct that the reviewers need to checkmark each box in the very first post before we move to the next step? Thanks!

**Comment by evanleeturner** [comment]
So I can't seem to edit the first post.  When I click on the ... my only options are "copy link" and "quote reply"

**Comment by danielskatz** [comment]
@whedon re-invite @evanleeturner as reviewer

**Comment by whedon** [comment]
The reviewer already has a pending invite.

@evanleeturner please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by evanleeturner** [comment]
I just did.  It said the invitation has expired.

**Comment by danielskatz** [comment]
👋 @evanleeturner - you shouldn't edit the first comment, but you should check boxes in it.  If you have comments, add them as new comments

**Comment by danielskatz** [comment]
@openjournals/dev - can you check on the issue @evanleeturner is having?

**Comment by evanleeturner** [comment]
It won't let me change the box when I click it - nothing happens.

**Comment by danielskatz** [comment]
that's a permission problem since you are not in the group - @openjournals/dev will need to fix it since the re-invite didn't work for some reason

**Comment by castelao** [comment]
@danielskatz , thanks!

**Comment by arfon** [comment]
Dear authors and reviewers

We wanted to notify you that in light of the current COVID-19 pandemic, JOSS has decided to suspend submission of new manuscripts and to handle existing manuscripts (such as this one) on a "best efforts basis". We understand that you may need to attend to more pressing issues than completing a review or updating a repository in response to a review. If this is the case, a quick note indicating that you need to put a "pause" on your involvement with a review would be appreciated but is not required.

Thanks in advance for your understanding.

_Arfon Smith, Editor in Chief, on behalf of the JOSS editorial team._

**Comment by arfon** [comment]
@whedon re-invite @evanleeturner as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@evanleeturner please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by danielskatz** [comment]
note that I tried that a few hours ago and it didn't seem to work...

**Comment by arfon** [comment]
Whedon's invite functionality is definitely making the invite correctly:

<img width="544" alt="Screen Shot 2020-03-13 at 8 29 39 PM" src="https://user-images.githubusercontent.com/4483/76670933-c5a33380-6569-11ea-9f07-75702a4ab9cb.png">

@evanleeturner - are you sure you're logged in with the correct GitHub account when clicking the link?

**Comment by jessicaaustin** [comment]
@castelao 

> 
> Thanks for sharing your opinion on using xarray or pandas. It’s now clear for me that I need to clarify this in the documentation. This will be addressed at https://cotede.readthedocs.io/en/dev/data_model.html
> 
> In summary, I believe that the data model adopted by CoTeDe does not limit the users, but the other way around: it gives more freedom. The package xarray, which was initially based on the netCDF structure, satisfies the data model expected by CoTeDe, thus most of CoTeDe’s functionalities can be used without any change, [see the example](https://github.com/castelao/CoTeDe/blob/master/docs/notebooks/TSG_Saildrone.ipynb). Since xarray can read directly netCDF and pandas objects can be converted into xarray objects, all those inputs can be used by CoTeDe, plus any other format that satisfies CoTeDe’s minimalist data model. CoTeDe has been used with SQL databases and MatLab based pipelines, just two examples where less requirements makes a difference.

Yes, I agree that CoTeDe's model is flexible and I don't think you should change it. All I am saying is that many people will come into this with their data model already in xarray/pandas/netcdf/etc and having some example code to copy/paste to get them started quickly will be a huge help. Probably some unit tests too, to be safe. Basically, since those data models are widely used in the python community, it should be seamless to use CoTeDe if your data is already in that format.  Sorry I don't think I made this clear in my initial review.


> > * **Highly recommended:** It would be great to list similar automated QC packages and describe how CoTeDe compares. For example, CoTeDe is more generic than ioos_qc, which depends on pandas/xarray and assumes netCDF integration, and CoTeDe allows integration with a wide range of QC test suites, whereas ioos_qc (currently) only focuses on QARTOD.
> >   * I do understand that this takes time, so I don't want it to block acceptance here. But this is the first question I had when I looked at this library, and would love to see it in the README someday.
> > ```
> 
> I see your point and I will consider that. The brief comparison that you suggested above could be a good start, but what should be said about the similarities between the two packages? For instance, what should we say about QC configuration, which seems like ioss_qc followed a quite similar approach as the one proposed by CoTeDe. Would you agree? On a side note, since this is a highly recommended suggestion, I was expecting that ioos_qc would adopt the same policy, but I could not find any reference to CoTeDe in the ioos_qc documentation.

Yes, like I said this is an involved process and can not be done quickly. Maybe you could put a placeholder in your documentation and fill it in as you (or other contributors) can get to it? Even just a list of other packages you think are comparable would be useful. 

You're right, ioos_qc does not do this, but one of the review points above for submission to JOSS mentions, "State of the field: Do the authors describe how this software compares to other commonly-used packages?" so that's where this is coming from. 

> > Documentation:
> > ```
> > * **BLOCKER:** The important classes and most of the test methods have docstrings, which is great. But I was not able to find any generated API docs under https://cotede.readthedocs.io/ . [castelao/CoTeDe#41](https://github.com/castelao/CoTeDe/issues/41)
> > ```
> 
> I have expanded the documentation (https://cotede.readthedocs.io/) to include the API reference for the public resources, i.e. everything that a new user would need to know to use CoTeDe. I’m also expanding the notebooks with practical examples to better illustrate how to use the package. I intend to extend the API reference with internal resources that might be useful for developers (https://cotede.readthedocs.io/en/latest/api.html).

https://cotede.readthedocs.io/en/latest/api.html does not work for me. Is this still in progress?

> > * Highly recommended you add performance tests. During development of https://github.com/ioos/ioos_qc we found that small tweaks to our testing algorithms could lead to severely degraded performance on large datasets. Recommend you add these tests to the travis-ci build and add it to the checklist of items to review if someone submits a PR.
> > ```
> 
> Thank you, that is a very good idea. I've always assumed that by using Python it was implied that speed was not a priority

IMHO people aren't going to be too worried about a fraction of second difference, but if it suddenly goes from <1s to evaluate a dataset to 20s, you will want to know what commit caused that regression. This is exactly the scenario that happened with ioos_qc, and it's this sort of catastrophic issue that I'm talking about.

**Comment by castelao** [comment]
@jessicaaustin , thanks for the fast response. Yes, my bad. The latest version of the documentation is on the dev branch:

https://cotede.readthedocs.io/en/dev/api.html

**Comment by jessicaaustin** [comment]
Ok great thanks. I'm going to check off the rest of the items on my checklist; at this point I think this is ready to accept.

**Comment by evanleeturner** [comment]
Looks like i was able to check off the first items...

I am ready to accept this paper!

**Comment by evanleeturner** [comment]
Thanks everyone so much for your patience!  Everything is crazy crazy here for Covid-19 and I'm working remote today.. so want to get this accepted and done while I am online!

**Comment by arfon** [comment]
> Looks like i was able to check off the first items...

@evanleeturner - thanks for the update. Just verifying that you've completed your review? If so, could you update the checklist accordingly (if permissions are fixed for you now).

**Comment by evanleeturner** [comment]
Yes, I am finished. When I click one box it sits there and locks the others - so just keeping diligent here and checking all of them as they open up for me.

**Comment by kthyng** [comment]
> 
> > ```
> > * Highly recommended, if possible: Include a list of groups actively using this library, to give us an idea of how widely used it is. (this could also be added to the code documentation)
> > ```
> 
> Thanks, that’s a good idea. However, I believe that the choice to be listed belongs to the users. I’ll create a designated space for that on CoTeDe’s website and encourage the users to add themselves to the list.
> 

As listed on the [JOSS website](https://joss.readthedocs.io/en/latest/submitting.html), authors should in their paper: "Mention (if applicable) a representative set of past or ongoing research projects using the software and recent scholarly publications enabled by it."

**Comment by kthyng** [comment]
@castelao I see that the boxes are checked off and both reviewers have stated that they are ready to accept. Please address my note immediately above. 

Also, there are two PRs that are open above. Can you and @jessicaaustin state if there are any lingering issues associated with them?

**Comment by jessicaaustin** [comment]
So long as https://github.com/castelao/CoTeDe/issues/41 is resolved and those API docs are available at https://cotede.readthedocs.io/en/latest/ , I'm ok with accepting. My other PRs/issues are nice-to-have.

**Comment by castelao** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02063/joss.02063/10.21105.joss.02063.pdf)

**Comment by castelao** [comment]
Thanks to everyone involved in this process, I do appreciate your time on this.

@kthyng I believe that I addressed everything. What is the next step? Shall I do a new release and obtain a new DOI? Thanks,

**Comment by kthyng** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.13155/33951 is OK
- 10.1016/j.mio.2014.09.001 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by kthyng** [comment]
@castelao Your references need some tweaks. I put some in [here](https://github.com/castelao/CoTeDe/pull/44) which you can merge if you approve. However, I see that the "note" category from "manual" entries is not coming through. Can you play around with these entries to get that information in?

**Comment by castelao** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.13155/33951 is OK
- 10.1016/j.mio.2014.09.001 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by castelao** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02063/joss.02063/10.21105.joss.02063.pdf)

**Comment by castelao** [comment]
@kthyng I'm sorry for my lack of attention on the bibliography records and thank you very much for your PR. I did a few other adjustments.

**Comment by castelao** [comment]
I just release a new version of CoTeDe and the new DOI is: 10.5281/zenodo.583710

Do I need to do anything else? Thank you

**Comment by kthyng** [comment]
@castelao I'm still seeing that the "note" section in your .bib file isn't coming through to be printed in your paper. For example, in the .bib file for Boyer et al, there is a note for the Technical Editor, but this doesn't come through to the reference in your paper. I can try to work on this at some point to see which entry type will carry through the note field, but it might be a few days before I have a chance.

**Comment by castelao** [comment]
@kthyng, you are already too generous with your time, I can do it myself.    Thank you.

Yes, I noticed that, but my understanding is that the lack of the note field is the result of the reference type 'manual' combined with the bibliography style used by JOSS. Different journals adopt different styles, and since this is not my expertise, I always provide the fields and let each journal style to decide how to use those fields.

One alternative is to move back from manual to incollection, as it was originally, which would allow an editor field. But notice that Mishonov is a technical editor and not a regular editor, so it would not be fully correct. My opinion is that hacking it to force those changes would go against the principle of using BibTeX. I'm more inclined to provide the fields and let the style definition decides how to use those and what to show. The fields currently shown are enough to uniquely identify this reference, and it is normal to have fields that are not shown in the reference. For instance, most styles omit the fields 'keys' and 'URL'.

If you think that it is necessary to show the technical editor, I can figure this out.

**Comment by kthyng** [comment]
@castelao Sure, the combination of items results in what ultimately shows up in the references. The thing that is most in your control is what type of entry a given reference is entered as. I think it is a good goal to have the information about the references show up, and it doesn't taking Latex hacking so much as appropriate field/entry choice. When I looked up that reference online, it was requested to show the Technical Editor in the citation, so I would try to respect that. That wasn't the only entry that had information in the bibtex entry that wasn't showing up in the references list. Can you give it another small chunk of time to do your best to faithfully represent the references?

**Comment by castelao** [comment]
@kthyng certainly. I just did a test, and if I replace the note field by an editor field it shows up in the PDF. The problem is that the editor field is not standard for a manual type of entry. But editor field is standard for an 'incollection' entry.

Was intentional that you changed [from 'incollection' to 'manual'](https://github.com/castelao/CoTeDe/commit/bd03f3369ab430f19a6654525ba0f0700f8863ca)? I could move it back to incollection, or I can exchange 'note' by 'editor' and keep the 'manual' type of entry. Let me know your preference.

**Comment by kthyng** [comment]
@castelao I changed from incollection to manual when I saw your other bib entries were manual and had note fields — I assumed that meant the note fields were coming through. And as I recall, the editor field wrote "Editor" but since the person is a technical editor, not just editor, I figured a less specific field would be more appropriate to use (then keeping the language "Technical Editor").

**Comment by castelao** [comment]
One alternative to force that output is to use:

...
series = {(A. V. Mishonov, {T}echnical {E}ditor) {NOAA} {Atlas} {NESDIS}
87},
...

Would you prefer that or incollection with editor (which would be 'Ed.'
only instead of Technical Editor)?

**Comment by castelao** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02063/joss.02063/10.21105.joss.02063.pdf)

**Comment by castelao** [comment]
@kthyng I spoke with Tim Boyer today, the first author of that publication, and he said that on his side there is no problem with using 'Editor' instead of 'Technical Editor'.

**Comment by kthyng** [comment]
ok great, let's move ahead then.

**Comment by kthyng** [comment]
@whedon add 10.5281/zenodo.583710 as archive

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by kthyng** [comment]
@whedon set 10.5281/zenodo.583710 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.583710" target="_blank">10.5281/zenodo.583710</a> is the archive.

**Comment by kthyng** [comment]
@castelao Can you edit the metadata at your Zenodo archive so that the title and authors on your paper exactly match there?

Also, what is the proper version number for your repository currently?

**Comment by castelao** [comment]
@kthyng I just updated the DOI record itself, thanks for noticing that.

The current version is 0.21.3

**Comment by kthyng** [comment]
@whedon set 10.5281/zenodo.3733959 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3733959" target="_blank">10.5281/zenodo.3733959</a> is the archive.

**Comment by kthyng** [comment]
Looks like the zenodo archive changed numbers too, so I updated that. Let me know if this is incorrect

**Comment by kthyng** [comment]
@whedon set v0.21.3 as version

**Comment by whedon** [comment]
OK. v0.21.3 is the version.

**Comment by kthyng** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.13155/33951 is OK
- 10.1016/j.mio.2014.09.001 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1410

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1410, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kthyng** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1247595390389583878) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1411
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02063](https://doi.org/10.21105/joss.02063)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by kthyng** [comment]
Congratulations to @castelao on your new publication! Many thanks to reviewers @jessicaaustin and @evanleeturner — we have relied on your time and expertise!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02063/status.svg)](https://doi.org/10.21105/joss.02063)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02063">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02063/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02063/status.svg
   :target: https://doi.org/10.21105/joss.02063
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02063/status.svg)](https://doi.org/10.21105/joss.02063)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
